styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesGCPServiceAccountSecretManager.java

```diff
@@ -0,0 +1,269 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.docker;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.iam.v1.model.ServiceAccountKey;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.hash.Hashing;
+import com.spotify.styx.ServiceAccountKeyManager;
+import com.spotify.styx.util.GcpUtil;
+import io.fabric8.kubernetes.api.model.PodList;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import io.fabric8.kubernetes.api.model.SecretList;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.NamespacedKubernetesClient;
+import java.io.IOException;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class KubernetesGCPServiceAccountSecretManager  {
+
+  private static final Logger logger = LoggerFactory.getLogger(KubernetesGCPServiceAccountSecretManager.class);
+
+  private static final String STYX_WORKFLOW_SA_ID_ANNOTATION = "styx-wf-sa";
+  private static final String STYX_WORKFLOW_SA_EPOCH_ANNOTATION = "styx-wf-sa-epoch";
+  private static final String STYX_WORKFLOW_SA_JSON_KEY_NAME_ANNOTATION = "styx-wf-sa-json-key-name";
+  private static final String STYX_WORKFLOW_SA_P12_KEY_NAME_ANNOTATION = "styx-wf-sa-p12-key-name";
+  private static final String STYX_WORKFLOW_SA_SECRET_NAME = "styx-wf-sa-keys";
+  private static final String STYX_WORKFLOW_SA_JSON_KEY = "styx-wf-sa.json";
+  private static final String STYX_WORKFLOW_SA_P12_KEY = "styx-wf-sa.p12";
+
+  private static final EpochProvider DEFAULT_SECRET_EPOCH_PROVIDER =
+      KubernetesGCPServiceAccountSecretManager::smearedDailyEpoch;
+
+  private static final Clock DEFAULT_CLOCK = Clock.systemUTC();
+
+  private static final Duration SECRET_GC_GRACE_PERIOD = Duration.ofMinutes(30);
+
+  private final KubernetesClient client;
+  private final ServiceAccountKeyManager serviceAccountKeyManager;
+  private final EpochProvider epochProvider;
+  private final Clock clock;
+
+  private final Object secretMutationLock = new Object() {};
+
+  KubernetesGCPServiceAccountSecretManager(
+      NamespacedKubernetesClient client,
+      ServiceAccountKeyManager serviceAccountKeyManager,
+      EpochProvider epochProvider,
+      Clock clock) {
+    this.client = Objects.requireNonNull(client).inNamespace("default");
```

Comment:
Maybe it should be set by StyxScheduler passing the k8s client to the runner and secret manager?